File: app/src/main/java/com/mmjang/ankihelper/data/history/HistoryUtil.java
package com.mmjang.ankihelper.data.history;

import android.content.Context;

import com.mmjang.ankihelper.data.database.AppDatabase;

/**
 * Utility class for saving history entries using Room database
 */
public class HistoryUtil {

    private static HistoryRepositoryHelper repositoryHelper;

    /**
     * Initialize the utility with application context
     * Must be called before any save methods
     */
    public static void initialize(Context context) {
        if (repositoryHelper == null) {
            AppDatabase database = AppDatabase.Companion.getInstance(context.getApplicationContext());
            HistoryRepository repository = new HistoryRepository(database.historyDao());
            repositoryHelper = new HistoryRepositoryHelper(repository);
        }
    }

    /**
     * Convert HistoryPOJO to HistoryEntity
     */
    private static HistoryEntity convertToEntity(HistoryPOJO pojo) {
        HistoryEntity entity = new HistoryEntity();
        entity.setTimeStamp(pojo.getTimeStamp());
        entity.setType(pojo.getType());
        entity.setWord(pojo.getWord());
        entity.setSentence(pojo.getSentence());
        entity.setDictionary(pojo.getDictionary());
        entity.setDefinition(pojo.getDefinition());
        entity.setTranslation(pojo.getTranslation());
        entity.setNote(pojo.getNote());
        entity.setTag(pojo.getTag());
        return entity;
    }

    public static void savePopupOpen(String sentence){
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.");
            return;
        }
        HistoryPOJO history = new HistoryPOJO();
        history.setType(HistoryType.POPUP_OPEN);
        history.setTimeStamp(System.currentTimeMillis());
        history.setSentence(sentence);
        repositoryHelper.insertHistory(convertToEntity(history));
    }

    public static void saveWordlookup(String sentence, String word){
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.");
            return;
        }
        HistoryPOJO history = new HistoryPOJO();
        history.setType(HistoryType.WORD_LOOK_UP);
        history.setTimeStamp(System.currentTimeMillis());
        history.setSentence(sentence);
        history.setWord(word);
        repositoryHelper.insertHistory(convertToEntity(history));
    }

    public static void saveNoteAdd(String sentence, String word,
                                   String dictionary, String definition,
                                   String translation, String note, String tag){
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.");
            return;
        }
        HistoryPOJO history = new HistoryPOJO();
        history.setType(HistoryType.NOTE_ADD);
        history.setTimeStamp(System.currentTimeMillis());
        history.setSentence(sentence);
        history.setWord(word);
        history.setDictionary(dictionary);
        history.setDefinition(definition);
        history.setTranslation(translation);
        history.setNote(note);
        history.setTag(tag);
        repositoryHelper.insertHistory(convertToEntity(history));
    }
}

File: app/src/main/java/com/lmyby/ankihelper/data/history/HistoryUtil.kt
package com.lmyby.ankihelper.data.history

import android.content.Context
import com.lmyby.ankihelper.data.database.AppDatabase

/**
 * Utility class for saving history entries using Room database
 * Converted to Kotlin as part of Phase 2 data model migration
 */
object HistoryUtil {
    private var repositoryHelper: HistoryRepositoryHelper? = null

    /**
     * Initialize the utility with application context
     * Must be called before any save methods
     */
    @JvmStatic
    fun initialize(context: Context) {
        if (repositoryHelper == null) {
            val database = AppDatabase.getInstance(context.applicationContext)
            val repository = HistoryRepository(database.historyDao())
            repositoryHelper = HistoryRepositoryHelper(repository)
        }
    }

    /**
     * Convert HistoryPOJO to HistoryEntity
     */
    private fun convertToEntity(pojo: HistoryPOJO): HistoryEntity {
        return HistoryEntity(
            timeStamp = pojo.timeStamp,
            type = pojo.type,
            word = pojo.word,
            sentence = pojo.sentence,
            dictionary = pojo.dictionary,
            definition = pojo.definition,
            translation = pojo.translation,
            note = pojo.note,
            tag = pojo.tag
        )
    }

    @JvmStatic
    fun savePopupOpen(sentence: String) {
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.")
            return
        }
        val history = HistoryPOJO(
            type = HistoryType.POPUP_OPEN,
            timeStamp = System.currentTimeMillis(),
            sentence = sentence
        )
        repositoryHelper!!.insertHistory(convertToEntity(history))
    }

    @JvmStatic
    fun saveWordlookup(sentence: String, word: String) {
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.")
            return
        }
        val history = HistoryPOJO(
            type = HistoryType.WORD_LOOK_UP,
            timeStamp = System.currentTimeMillis(),
            sentence = sentence,
            word = word
        )
        repositoryHelper!!.insertHistory(convertToEntity(history))
    }

    @JvmStatic
    fun saveNoteAdd(
        sentence: String,
        word: String,
        dictionary: String,
        definition: String,
        translation: String,
        note: String,
        tag: String
    ) {
        if (repositoryHelper == null) {
            android.util.Log.e("HistoryUtil", "HistoryUtil not initialized! Call initialize() first.")
            return
        }
        val history = HistoryPOJO(
            type = HistoryType.NOTE_ADD,
            timeStamp = System.currentTimeMillis(),
            sentence = sentence,
            word = word,
            dictionary = dictionary,
            definition = definition,
            translation = translation,
            note = note,
            tag = tag
        )
        repositoryHelper!!.insertHistory(convertToEntity(history))
    }
}
